pkg/operator/events/recorder_upstream.go

```diff
@@ -59,12 +125,33 @@ func (r *upstreamRecorder) Warningf(reason, messageFmt string, args ...interface
 	r.Warning(reason, fmt.Sprintf(messageFmt, args...))
 }
 
+func (r *upstreamRecorder) incrementEventsCounter(severity string) {
+	if r.involvedObjectRef == nil {
+		return
+	}
+	eventsCounterMetric.WithLabelValues(severity).Inc()
+}
+
 // Event emits the normal type event.
 func (r *upstreamRecorder) Event(reason, message string) {
+	r.shutdownMutex.Lock()
+	defer r.shutdownMutex.Unlock()
```

Comment:
this is a long lock. Use a RW lock and lock here read-only, and write-lock in the shutdown call func.